clippy_lints/src/eta_reduction.rs

```diff
@@ -145,6 +146,7 @@ impl<'tcx> LateLintPass<'tcx> for EtaReduction {
         );
 
         if_chain!(
+            if !is_adjusted(cx, &body.value);
```

Comment:
I think that `is_adjusted` is a bit too restrictive. The example cases all are related to the never type. Therefore, I would only suppress the lint if the wrapped function returns the never type, in all other cases it seems like the normal adjustment has worked its magic.

Have you tried the following two options?
1. Check the return type of the body expression of the closure. The expression might still return the never type if it gets adjusted, but I'm not sure about that.
* Otherwise we might have to resolve the referenced function and check what that returns.

Reply:
> The example cases all are related to the never type.

The problem is not `never` type, but it's when the wrapped function can only be resolved by typechecks by adjusting it to return `()` type.  

> Check the return type of the body expression of the closure. The expression might still return the never type if it gets adjusted, but I'm not sure about that.

It gets adjusted to `()` type.

> Otherwise we might have to resolve the referenced function and check what that returns.

Yes, the only way that I see for a complete solution is to check if the function typecheck can be resolved to accept the original return type of the closure. I'm not sure this is trivial to do.

Reply:
> > The example cases all are related to the never type.
> 
> The problem is not `never` type, but it's when the wrapped function can only be resolved by typechecks by adjusting it to return `()` type.

Looks like I also misunderstood the issue at first. The type adjustment to `()` sounds like a quirk of the type checking system.

In that case, I think that your fix is the correct one. It will have some FNs but those are better than FPs IMO. :upside_down_face: